core/src/main/java/org/infinispan/notifications/cachelistener/DistributedQueueingSegmentListener.java

```diff
@@ -77,12 +77,15 @@ public boolean handleEvent(EventWrapper<K, V, CacheEntryEvent<K, V>> wrapped, Li
 
    @Override
    public void transferComplete() {
-      justCompletedSegments.forEach(completeSegment);
-      completed.set(true);
-      notifiedKeys.clear();
+      // Complete any segments that are still open - means we just didn't receive the completion yet
```

Comment:
I'm having some trouble parsing this comment, could you elaborate on why we `completeSegment()` may not have been called for some of the segments?

Reply:
The reason it wouldn't be completed is that the completion is done on a different thread. I will see if I can tweak it to be done in a slightly different order to provide more clear ordering semantics.